File: 0525-contiguous-array/0525-contiguous-array.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        gMax = 0
        hmap = {}
        hmap[0] = 0
        pSum = 0
        
        for i in range(len(nums)):            
            if nums[i] == 1:
                pSum += 1 
            else:
                pSum -= 1
                
            if pSum in hmap:
                gMax = max(gMax, i+1 - hmap[pSum]) 
            
            if pSum not in hmap:
                hmap[pSum] = i + 1
        
        return gMax

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
      
        int pSum = 0;
        int gMax = 0;
        
        unordered_map<int, int> hmap;
        hmap[0] = 0;
        
        for( int i = 0; i < nums.size(); i++) {
            if (nums[i] == 1)
                pSum += 1;
            else
                pSum -= 1;
        
            // find the maximum array size of we have a sub array of 
            // unequal 0's and 1's with minimum size
            if (hmap.find(pSum) != hmap.end()) {
                gMax = std::max(gMax, ((i+1) - hmap[pSum]) );
            }
        
            // Add pSum for the current index if it does not already exist in hmap
            if (hmap.find(pSum) == hmap.end()) {
                hmap[pSum] = i + 1;
            }
        } 
        
        return gMax;
    }
};